### PR TITLE
ST: bridge integration

### PR DESCRIPTION
**Pull request by Frawless**
### Type of change

- Enhancement / new feature

### Description

Integration of [Strimzi Bridge](https://github.com/strimzi/strimzi-kafka-bridge) into Strimzi test suite.

### Checklist

- [x] Write tests
- [ ] Make sure all tests pass
- [ ] Update documentation

### CODE REVIEW

**Comment by scholzj** [comment]
I wonder if this should be really integrated here and the way it is done here: 
* The Bridge has a separate repo and will work also separately outside of OpenShift. So maybe the system tests should be there?
* When finished, it will be integrated into the Operator through its own Custom resource - not through deploying it as an image.

**Comment by Frawless** [comment]
@scholzj I agree, but here we have all things rdy for deploy kafka and other necessary things. I am not sure if we will be able to create some suite in bridge repo till next release (I will be out for two weeks). Here is a lot of things to discuss from my POV, but right now, this way is the easiest for us to have at least some automated coverage without minimal code.

In case we will create some suite in bridge repo, how we will manage cluster deployment in openshift? Keep tests in bridge repo and use deployment from operator repo?

**Comment by scholzj** [comment]
> In case we will create some suite in bridge repo, how we will manage cluster deployment in openshift? Keep tests in bridge repo and use deployment from operator repo?

The bridge should be available as a standalone thing (e.g. on RHEL) as well as through the Cluster Operator. I do not think we have sorted out the details. But in case of the CO deployment, there will be probably some `KafkaHttpProxy` custom resource or something similar. However, I wonder if it would be easier to just test most fo the proxy outside of OpenShift with Kafka running locally and afterwards just test that it is properly deployed on Kubernetes. The Minikube / OpenShift in tests doesn't add on stability, just on complexity - that is why I'm not sure whether it wouldn't be better to test the bridge functionality separately.

**Comment by ppatierno** [comment]
I totally agree with @scholzj. The bridge ST should be in the related repo even because the bridge can run outside of OpenShift or Kubernetes but just alongside Kafka on bare metal.

**Comment by Frawless** [comment]
@strimzi-ci run tests testcase=HttpBridgeST

**Comment by Frawless** [comment]
@strimzi-ci run tests testcase=HttpBridgeST

File: systemtest/pom.xml [hunk 1]
```diff
@@ -122,19 +122,26 @@
         <dependency>
             <groupId>io.vertx</groupId>
             <artifactId>vertx-web-client</artifactId>
-            <version>${vertx.version}</version>
+            <version>3.7.0</version>
```
Comment (ppatierno) [inline systemtest/pom.xml]: do you need the specific 3.7.0 version for any reason? Vert.x Kafka client or something like that? Btw we are on 3.7.1 now.

File: systemtest/src/main/java/io/strimzi/systemtest/Constants.java [hunk 2]
```diff
@@ -43,6 +43,18 @@
     String STRIMZI_DEPLOYMENT_NAME = "strimzi-cluster-operator";
     String IMAGE_PULL_POLICY = "Always";
 
+    /**
+     * Kafka Bridge JSON encoding with JSON embedded format
+     */
+    String KAFKA_BRIDGE_JSON_JSON = "application/vnd.kafka.json.v2+json";
```
Comment (ppatierno) [inline systemtest/src/main/java/io/strimzi/systemtest/Constants.java]: the one with BINARY embedded format is missing.

File: systemtest/src/test/java/io/strimzi/systemtest/HttpBridgeST.java [hunk 3]
```diff
@@ -0,0 +1,279 @@
...
+                        future.complete(response.body());
+                    } else {
+                        LOGGER.info("Cannot consume any messages!");
+                        future.completeExceptionally(ar.cause());
+                    }
+                });
+        return future.get(1, TimeUnit.MINUTES);
+    }
+
+    boolean subscribHttpConsumer(JsonObject topics, String bridgeHost, int bridgePort, String groupId, String name) throws InterruptedException, ExecutionException, TimeoutException {
```
Comment (ppatierno) [inline systemtest/src/test/java/io/strimzi/systemtest/HttpBridgeST.java]: typo "subscrib"

**Comment by ppatierno** [comment]
Didn't we agree on having these ST in the bridge project? I see duplicated tests somehow between the two projects, or?

**Comment by Frawless** [comment]
@ppatierno yes we did, but currently we don't have support for re-use deployment features from operator tests. Basically, this test is mainly for the operator, if everything works fine after deployment by CO. In Bridge repo, we should focus on test with standalone kafka (debezium one ideally and have there option for run it against  kafka running somewhere and against cluster in openshift) I think. Adding tests which needs CO deployment in Bridge repo will be kinda hard to do even with current state of operator test suite.

**Comment by ppatierno** [comment]
@Frawless makes sense. So If I got it right, here we should expect just a few tests against the bridge to check that it was deployed in the right way by CO. While from a features point of view related to the bridge, all the tests will be in the bridge repo. Ok!

**Comment by Frawless** [comment]
@Frawless  yes, that's what I had in my mind :)

File: systemtest/pom.xml (hunk 1 above)
Comment (Frawless) [reply systemtest/pom.xml]: Changed to global version used in the project

File: systemtest/src/test/java/io/strimzi/systemtest/HttpBridgeST.java (hunk 3 above)
Comment (Frawless) [reply systemtest/src/test/java/io/strimzi/systemtest/HttpBridgeST.java]: fixed

**Comment by Frawless** [comment]
@strimzi-ci run tests testcase=HttpBridgeST,HttpBridgeTlsST

**Comment by Frawless** [comment]
@strimzi-ci run tests testcase=HttpBridgeST,HttpBridgeTlsST

**Comment by Frawless** [comment]
@strimzi-ci run tests testcase=HttpBridgeST,HttpBridgeTlsST,HttpBridgeScramShaST

**Comment by Frawless** [comment]
@strimzi-ci run tests testcase=HttpBridgeST,HttpBridgeTlsST,HttpBridgeScramShaST

**Comment by Frawless** [comment]
@strimzi-ci run tests testcase=HttpBridgeST,HttpBridgeTlsST,HttpBridgeScramShaST

**Comment by Frawless** [comment]
@strimzi-ci run tests testcase=HttpBridgeST,HttpBridgeTlsST,HttpBridgeScramShaST

**Comment by Frawless** [comment]
@strimzi-ci run tests testcase=HttpBridgeST,HttpBridgeTlsST,HttpBridgeScramShaST

**Comment by Frawless** [comment]
@strimzi-ci run tests testcase=HttpBridgeST,HttpBridgeTlsST,HttpBridgeScramShaST

**Comment by Frawless** [comment]
@strimzi-ci run tests testcase=HttpBridgeST,HttpBridgeTlsST,HttpBridgeScramShaST

**Comment by Frawless** [comment]
@strimzi-ci run tests testcase=HttpBridgeST,HttpBridgeTlsST,HttpBridgeScramShaST

**Comment by Frawless** [comment]
@strimzi-ci run tests testcase=HttpBridgeST,HttpBridgeTlsST,HttpBridgeScramShaST

**Comment by Frawless** [comment]
@strimzi-ci run tests testcase=HttpBridgeST,HttpBridgeTlsST,HttpBridgeScramShaST

File: systemtest/src/main/java/io/strimzi/systemtest/Environment.java [hunk 4]
```diff
@@ -30,6 +30,10 @@
      * Specify test-client image used in system tests.
      */
     private static final String TEST_CLIENT_IMAGE_ENV = "TEST_CLIENT_IMAGE";
+    /**
+     * Specify test-client image used in system tests.
```
Comment (ppatierno) [inline systemtest/src/main/java/io/strimzi/systemtest/Environment.java]: copy/paste error :P it should be "bridge image"

File: systemtest/src/main/java/io/strimzi/systemtest/Environment.java [hunk 5]
```diff
@@ -61,7 +65,7 @@
     private static final String STRIMZI_LOG_LEVEL_DEFAULT = "DEBUG";
     static final String KUBERNETES_DOMAIN_DEFAULT = ".nip.io";
     private static final String STRIMZI_FULL_RECONCILIATION_INTERVAL_MS_DEFAULT = "30000";
-    static final int INGRESS_DEFAULT_PORT = 4242;
+    public static final int KAFKA_CLIENTS_DEFAULT_PORT = 4242;
```
Comment (ppatierno) [inline systemtest/src/main/java/io/strimzi/systemtest/Environment.java]: out of curiosity, why this port choice?

File: systemtest/src/main/java/io/strimzi/systemtest/Environment.java [hunk 6]
```diff
@@ -75,6 +79,9 @@
     // variables for test-client image
     private static final String TEST_CLIENT_IMAGE_DEFAULT = STRIMZI_REGISTRY + "/" + STRIMZI_ORG + "/test-client:" + STRIMZI_TAG + "-kafka-" + ST_KAFKA_VERSION;
     public static final String TEST_CLIENT_IMAGE = System.getenv().getOrDefault(TEST_CLIENT_IMAGE_ENV, TEST_CLIENT_IMAGE_DEFAULT);
+    // variables for test-client image
```
Comment (ppatierno) [inline systemtest/src/main/java/io/strimzi/systemtest/Environment.java]: "for bridge image"

File: systemtest/src/main/java/io/strimzi/systemtest/HttpBridgeBaseST.java [hunk 7]
```diff
@@ -0,0 +1,227 @@
...
+
+    protected String brdigeExternalService = CLUSTER_NAME + "-bridge-external-service";
+
+    @BeforeAll
+    void prepareEnv(Vertx vertx) {
+        // Create http client
+        client = WebClient.create(vertx, new WebClientOptions()
+                .setSsl(false)
+                .setTrustAll(true)
+                .setVerifyHost(false));
```
Comment (ppatierno) [inline systemtest/src/main/java/io/strimzi/systemtest/HttpBridgeBaseST.java]: why setting these properties related to SSL if you are using SSL=false so verify host and trust all aren't needed?

Comment (Frawless) [reply systemtest/src/main/java/io/strimzi/systemtest/HttpBridgeBaseST.java]: TBH I am not sure where I get it, I will remove it and push fixes for other typos

File: systemtest/src/main/java/io/strimzi/systemtest/Constants.java (hunk 2 above)
Comment (serrss) [reply systemtest/src/main/java/io/strimzi/systemtest/Constants.java]: @Frawless I guess Paolo talking about `"application/vnd.kafka.binary.v2+json"`

**Pull request merged**